video/out/wayland_common.c

```diff
@@ -2166,6 +2166,7 @@ bool vo_wayland_init(struct vo *vo)
     update_app_id(wl);
     mp_make_wakeup_pipe(wl->wakeup_pipe);
 
+    wl->callback_surface = wl->surface;
     wl->frame_callback = wl_surface_frame(wl->surface);
     wl_callback_add_listener(wl->frame_callback, &frame_listener, wl);
     wl_surface_commit(wl->surface);
```

Comment:
This does work but I don't like how it the "wrong" surface is picked at first for `vo_dmabuf_wayland` and it corrects itself later. This is kind of expanding the scope but I think it's worth it to add a bool in the `vo_wayland_state` that stores if the vo driver is in fact `dmabuf-wayland`. Then we can use that to make this decision instead.

Reply:
I think that boolean state is already known by virtue of not having any buffer attached to the child surface, but I'll have a look...

Reply:
@Dudemanguy, it doesn't look that simple to change. `vo_wayland_init` both creates the `vo_wayland_state` and the first *wrong* frame callback in a single function. `vo_dmabuf_wayland` can't change that unless:
  * A parameter is added to `vo_wayland_init`; or
  * The first `wl_surface_frame` is moved out of `vo_wayland_init`.
 
So unless I've misunderstood something, the current design here feels cleanest, despite your concern which I also agree with.

Reply:
You can add that parameter in `vo_wayland_init`. `vo_dmabuf_wayland` always calls that function first (it's done via `ra_ctx_create_by_name`).

Something like:
```
wl->dmabuf_wayland = strcmp(wl->vo->driver->name, "dmabuf-wayland") == 0;
wl->callback_surface = wl->dmabuf_wayland ? wl->video_surface : wl->surface;
wl->frame_callback = wl_surface_frame(wl->callback_surface);
```

In `vo_wayland_init` would do the trick. Change the other parts as needed of course.

Reply:
That's not a parameter, that's a variable. Also not ideal to embed knowledge of `vo_dmabuf_wayland` into `wayland_common.c`.

Instead, a parameter would look something like this:
```c
bool vo_wayland_init(struct vo *vo, bool use_video_surface);
```

Reply:
>Also not ideal to embed knowledge of vo_dmabuf_wayland into wayland_common.c

It already is by the virtue of having `vo` being a member of the struct. This is just a shorthand. dmabuf_wayland detection is done multiple times already actually.

Reply:
Oh I see. Yes
```c
bool dmabuf_wayland = !strcmp(wl->vo->driver->name, "dmabuf-wayland");
```
is already used twice in `wayland_common.c`. Not a design I'd recommend but I'm happy to not be introducing anything inconsistent by doing it a third time.

Reply:
Done.